frontend/css_and_js.py

```diff
@@ -1,37 +1,32 @@
 from os import path
 
+
 def readTextFile(*args):
     dir = path.dirname(__file__)
     entry = path.join(dir, *args)
     with open(entry, "r", encoding="utf8") as f:
         data = f.read()
     return data
 
+
 def css(opt):
     styling = readTextFile("css", "styles.css")
     if not opt.no_progressbar_hiding:
         styling += readTextFile("css", "no_progress_bar.css")
     return styling
 
+
 def js(opt):
     data = readTextFile("js", "index.js")
     data = "(z) => {" + data + "; return z ?? [] }"
     return data
 
-# Wrap the typical SD method call into async closure for ease of use
-# If you call frontend method without wrapping
-# DONT FORGET to bind input argument if you need it: SD.with(x)
-def w(sd_method_call):
-    return f"async (x) => {{ return await SD.with(x).{sd_method_call} ?? x ?? []; }}"
-
-def js_move_image(from_id, to_id):
-    return w(f"moveImageFromGallery('{from_id}', '{to_id}')")
-
-def js_copy_to_clipboard(from_id):
-    return w(f"copyImageFromGalleryToClipboard('{from_id}')")
 
-def js_painterro_launch(to_id):
-    return w(f"Painterro.init('{to_id}')")
-
-def js_img2img_submit(prompt_row_id):
-    return w(f"clickFirstVisibleButton('{prompt_row_id}')")
+# Wrap the typical SD method call into async closure for ease of use
+# Supplies the js function with a params object
+# That includes all the passed arguments and input from Gradio: x
+def call_SD(sd_method, **kwargs):
+    if "x" not in kwargs.keys():
+        kwargs["x"] = "x"
+    params = "{" + ",".join(f"{k}:{v}" for k, v in kwargs.items()) + "}"
```

Comment:
I wonder if there's some JSON formatter we could use here instead to get character escaping automatically? This is probably fine for now though

Reply:
Copying this from the other comment chain:

I would do recommend doing it like this:

```py
import json

some_options = {
  "num": 123,
  "bool": True,
  "str": "this is a string"
}

json_str = json.dumps(some_options)

function_name = "some_function"

function_call_str = f"async (x) => {{ return await SD.{function_name}({{ ...{json_str}, x }}) ?? []; }}"
```

which results in:

```js
async (x) => { return await SD.some_function({ ...{"num": 123, "bool": true, "str": "this is a string"}, x }) ?? [] }
```

which is essentially the same as this after the object spread is evaluated:

```js
async (x) => { return await SD.some_function({ num: 123, bool: true, str: "this is a string", x }) ?? [] }
```

And the original types are retained without needing to manually add quotes around strings or convert `True` to `true` etc.

Reply:
This was resolved by [4da7cb6](https://github.com/hlky/stable-diffusion-webui/pull/381/commits/4da7cb643723c8fb7c5f9f9a06353b709534dd79)